packages/interactivity/src/proxies/test/state-proxy.ts

```diff
@@ -0,0 +1,1187 @@
+				},
+				set double( val ) {
+					state.counter = val / 2;
+				},
+			} );
+			expect( state.counter ).toBe( 1 );
+			state.double = 4;
+			expect( state.counter ).toBe( 2 );
+		} );
+
+		it( 'should update array length', () => {
+			expect( state.array.length ).toBe( 2 );
+			state.array.push( 4 );
+			expect( state.array.length ).toBe( 3 );
+			state.array.splice( 1, 2 );
+			expect( state.array.length ).toBe( 1 );
+		} );
+
+		it( 'should update when mutations happen', () => {
+			expect( state.a ).toBe( 1 );
+			state.a = 11;
+			expect( state.a ).toBe( 11 );
+		} );
+
+		it( 'should support setting getters on the fly', () => {
+			const state = proxifyState< {
+				counter: number;
+				double?: number;
+			} >( 'test', {
+				counter: 1,
+			} );
+			Object.defineProperty( state, 'double', {
+				get() {
+					return state.counter * 2;
+				},
+			} );
+			expect( state.double ).toBe( 2 );
+			state.counter = 2;
+			expect( state.double ).toBe( 4 );
+		} );
+
+		it( 'should copy object like plain JavaScript', () => {
+			const state = proxifyState< {
+				a?: { id: number; nested: { id: number } };
+				b: { id: number; nested: { id: number } };
+			} >( 'test', {
+				b: { id: 1, nested: { id: 1 } },
+			} );
+
+			state.a = state.b;
+
+			expect( state.a.id ).toBe( 1 );
+			expect( state.b.id ).toBe( 1 );
+			expect( state.a.nested.id ).toBe( 1 );
+			expect( state.b.nested.id ).toBe( 1 );
+
+			state.a.id = 2;
+			state.a.nested.id = 2;
+			expect( state.a.id ).toBe( 2 );
+			expect( state.b.id ).toBe( 2 );
+			expect( state.a.nested.id ).toBe( 2 );
+			expect( state.b.nested.id ).toBe( 2 );
+
+			state.b.id = 3;
+			state.b.nested.id = 3;
+			expect( state.b.id ).toBe( 3 );
+			expect( state.a.id ).toBe( 3 );
+			expect( state.a.nested.id ).toBe( 3 );
+			expect( state.b.nested.id ).toBe( 3 );
+
+			state.a.id = 4;
+			state.a.nested.id = 4;
+			expect( state.a.id ).toBe( 4 );
+			expect( state.b.id ).toBe( 4 );
+			expect( state.a.nested.id ).toBe( 4 );
+			expect( state.b.nested.id ).toBe( 4 );
+		} );
+
+		it( 'should be able to reset values with Object.assign', () => {
+			const initialNested = { ...nested };
+			const initialState = { ...raw, nested: initialNested };
+			state.a = 2;
+			state.nested.b = 3;
+			Object.assign( state, initialState );
+			expect( state.a ).toBe( 1 );
+			expect( state.nested.b ).toBe( 2 );
+		} );
+
+		it( 'should keep assigned object references internally', () => {
+			const obj = {};
+			state.nested = obj;
+			expect( raw.nested ).toBe( obj );
+		} );
+
+		it( 'should keep object references across namespaces', () => {
+			const raw1 = { obj: {} };
+			const raw2 = { obj: {} };
+			const state1 = proxifyState( 'test-1', raw1 );
+			const state2 = proxifyState( 'test-2', raw2 );
+			state2.obj = state1.obj;
+			expect( state2.obj ).toBe( state1.obj );
+			expect( raw2.obj ).toBe( state1.obj );
+		} );
+
+		it( 'should use its namespace by default inside setters', () => {
```

Comment:
Why specifically within setters (and not within getters, for example)?

Reply:
I've added a test case for that in https://github.com/WordPress/gutenberg/pull/62734/commits/8a00d34d945fa32342a808f44da538af2ee7e0a2.